Fix: Reworking/improving optionator configuration for --print-config

platinumazure opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://github.com/eslint/eslint/blob/master/templates/bug-report.md))
[ ] New rule ([template](https://github.com/eslint/eslint/blob/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Changing behavior of a CLI option (--print-config).

<!--
    The following is required for all pull requests:
-->

**Please check each item to ensure your pull request is ready:**

- [x] I've read the [pull request guide](http://eslint.org/docs/developer-guide/contributing/pull-requests)
- [x] I've included tests for my change
- (n/a) I've updated documentation for my change (if appropriate)

**What changes did you make? (Give an overview)**

Since --print-config operates on one file, I've changed the option from being boolean-valued to being string-valued (in other words, the file for which user wants to print a config becomes an argument to the option itself).

~~This is technically a breaking change because this use is now broken: `eslint --print-config --other-options-here fileName.js`. In addition, it will no longer flag if too many files are passed in. However, the advantage of this is the optionator output for this option is improved and we get validation for free.~~ Consensus seems to be that this is not a breaking change, because `--print-config --some-option fileName.js` was never really intended to work, it just happened to because of the implementation.

Needed to update optionator to pick up a bugfix (otherwise, non-boolean options at the end of a command line were not validated properly).

**Is there anything you'd like reviewers to focus on?**

Nothing in particular.

ilyavolodin commented:
I don't think it's breaking, it's a bug fix. We always intended for `print-config` to work on a file, it was just not enforced before (bug).

nzakas commented on lib/cli.js:
```diff
@@ -137,6 +137,18 @@ const cli = {
 
             log.info("v" + require("../package.json").version);
 
+        } else if (currentOptions.printConfig) {
```
What happens if I do this:

```
$ eslint --print-config foo.js bar.js
```

In our current implementation, we get a warning. If I'm reading this code correctly, the new version would print the config for foo.js and not warn at all about bar.js, which I think is misleading. We probably want to check that `files.length` is zero, and if not, throw up the same warning we currently do.

platinumazure replied:
Yeah, I thought about that. You are right, the behavior has changed. So I can add that check back in. (I removed a unit test for that specific behavior. I guess I hadn't considered that use case important, oops.)

I'll add a check and I'll get that test back into tests/lib/cli.js. Sound good?

platinumazure commented:
@eslint/eslint-tsc Just want to confirm, should this be regarded as a bug-fix and non-breaking, per @ilyavolodin's comment [here](https://github.com/eslint/eslint/pull/7206#issuecomment-248793325)? Thanks!

platinumazure commented:
@nzakas I've added the logic and test back in, though I changed the error message for clarity. Please let me know if further changes are needed. Thanks!

nzakas commented on lib/options.js:
```diff
@@ -216,8 +216,8 @@ module.exports = optionator({
-            type: "Boolean",
-            description: "Print the configuration to be used"
+            type: "path::String",
+            description: "Print the configuration which would be used for linting a given file (linting will not occur)"
```
I'm guessing this will cause the text to wrap when output on the console. I'm not sure this needs to be changed from the previous text, but if it does, then can you make it a bit shorter?

platinumazure commented:
It does wrap, but prettily (I.e., first character of second line aligns
with first character of first line). I don't think it necessarily needs to
be shrunk, if you were thinking the wrapping would be ugly and go to the
left side of the console. Still want this shrunk?

On Sep 23, 2016 1:43 PM, "Nicholas C. Zakas" <notifications@github.com>
wrote:

> *@nzakas* requested changes on this pull request.
> ------------------------------
>
> In lib/options.js
> <https://github.com/eslint/eslint/pull/7206#pullrequestreview-1383483>:
>
> > @@ -216,8 +216,8 @@ module.exports = optionator({
>          },
>          {
>              option: "print-config",
> -            type: "Boolean",
> -            description: "Print the configuration to be used"
> +            type: "path::String",
> +            description: "Print the configuration which would be used for linting a given file (linting will not occur)"
>
> I'm guessing this will cause the text to wrap when output on the console.
> I'm not sure this needs to be changed from the previous text, but if it
> does, then can you make it a bit shorter?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/7206#pullrequestreview-1383483>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARWeiUyF5KgKb_3kbIKj1o0XX9l68VXks5qtB3CgaJpZM4KDXga>
> .
>

platinumazure commented:
Here's how my `eslint --help` looks with this change:

![eslint-help](https://cloud.githubusercontent.com/assets/284282/18805785/549194a8-81dc-11e6-8d0e-3ef6e65ccf88.png)

Personally I think this should work, but let me know if you want it shorter anyway.

nzakas commented:
I still think it should be shorter. We have documentation for details, these should just be short phrases.

platinumazure commented:
@nzakas Could I get a suggestion of what you're getting at? (Note that there are other options which go onto two lines.) I'm not sure how to shorten what I've got without losing meaning.

gkz commented:
Optionator does nice line wrapping at all widths. That being said, you could rephrase it to something like:

    Print the config to be used when linting the given file (linting won't occur)

nzakas commented:
@platinumazure I think you can just leave the message as it was before. I don't see a great need for changing it.

platinumazure commented:
@nzakas I've changed the option description-- it's shorter, but I wanted to at least allude to the option argument. It fits on one line on my terminal (which is 80 characters wide). I've also reclassified this as a non-breaking bugfix per consensus on this issue. Please let me know if anything else needs to change.
